### ISSUE TITLE
conflict between mat2x2 layout in uniform buffer and storage buffer

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#1258]
For mat2x2, the stride between columns is 16 bytes in In a uniform buffer and 8 bytes in storage buffer.
As part of the resolution to #922, we decided to avoid having the user tell us the matrix stride.

Then consider code like this:
```rust
[[block]]  struct S {
   [[offset(0)]]  M : mat2x2<f32>;
};

[[set(0), binding(1)]] var<storage> sbuf : S;
[[set(0), binding(1)]] var<storage> ubuf : S;
```

Then:
- ubuf.M[1][0] is 16 bytes into the buffer, and
- sbuf.M[1][0]  is 8 bytes into the buffer

This is ... confusing for users.

Also, when generating SPIR-V code for the type, we have to split S into two types, which causes a lot of complexity.

I suggest adding a rule:
- If a type contains a mat2x2, then that type may not be used as the store type or inside the store type for both a uniform buffer and a storage buffer.

We already have a note in [3.4.6.2. Internal Layout of Values](https://gpuweb.github.io/gpuweb/wgsl.html#internal-value-layout) that says:

> Note: Matrix values are laid out more compactly in the storage storage class than in the uniform storage class.

Thanks to @Kangz  who raised this to prominence.

**Comment by Kangz** [comment]
I'm a bit worried at the "validation at a distance" that would happen if we add the rule that `mat2x2` cannot be used in both uniform and storage.

It means that if I have a fairly common structure definition used in a bunch of places, (but only uniforms) and then I start to use it in a storage buffer, then things will fail validation. It might be better to have a `[[matrix_stride(N)]]` decoration so that if it is a multiple of `32`, then the matrix type is usable in both uniform and storage contexts.

**Comment by dneto0** [comment]
@alan-baker pointed out that for Vulkan, mat2x2 can be laid out tightly (MatrixStride of 8bytes).    That's tighter than what std140 allows.
It comes from the Vulkan rule  (emphasis added):

> A scalar, vector or **matrix type** has an extended alignment equal to its base alignment.


**So**.  I'll make a PR to eliminate this unusual difference.


To accomodate targeting OpenGL,  @Kangz  had a good idea.   mat2x2 is the only problematic case.   Its layout in a storage buffer has exactly the same footprint as a `vec4<f32>`. 

So:
-  WGSL can declare that `mat2x2<f32>`  is laid out in 4 consecutive 32-bit words. 
- When translating to OpenGL with the more restrictive layout, we materialize the matrix as if it were `vec4<f32>`
    - translate on load and translate on store.
    - there is one tricky bit which is you can form a pointer to one of the column vectors, and then load it or store it.  For that case, break it down into two loads and stores of the components.

**Comment by Kangz** [comment]
+1

Note that mat2x3 and mat2x4 (or 4x2 and 3x2) will also be issues in GLSL but we can hande them the same way.

**Comment by kainino0x** [comment]
Right now matrix strides are implicit. Someone suggested making them explicit so you wouldn't be able to get confused about the placement of the elements of the matrix. There would be only one valid value for the stride, so it would be pure "homework".

I pointed out that the stride is not ambiguous when the major dimension is 4 (`matXx4` I think). So we could have it implicit in that case, but required when the major dimension is 2 or 3 (`matXx2`, `matXx3`). We could even make it implicit when the major dimension is 3, as that's also a very common case due to `mat3x3 * vec3` (I'd expect commonness on par with 4, whereas 2 would be much less common).

**Comment by jdashg** [comment]
(We agree that mat2x2 should be tightly packed (like/as a vec4))

I would like for our show-your-work offset annotation requirements to protect me when I don't realize that webgpu is std140 *but* e.g. mat2x2 is tightly packed.

We had three approaches:
1. Forbid implicit padding, so subsequent offsets would have to acknowledge that mat2x2 is tightly packed
  * Still bug-prone if the mat2x2 is the last member
2. No spec change, but UAs can warn (from Corentin)
  * Users can eliminate the warning by using explicit padding members
  * Still bug-prone if the mat2x2 is the last member
3. Require show-your-work via matrix_stride annotations (from Dan)
  * We might only require it for the non-obvious strides, e.g. require for mat2x2, but not mat4x4.
  * This would prevent bugs even if mat2x2 is the last member

3 solves things most fully. I do sort of think implicit padding is a general footgun, so I also favor 1 in general. 2 very cleverly needs no spec change and makes me less worried about this whole issue in practice, though.

I think we should consider 3 for solving this issue. In particular it's weird that the actual implementation/layout of some struct depends on whether its bound via uniform or storage. By requiring matrix_stride for these ambiguous layouts, we entirely prevent this footgun.

**Comment by dneto0** [comment]
This is resolved except for demanding extra about oddball contextual layout cases, or requiring more annotations from the programmer.  See also  #1339 #1349.
Perhaps this can be consolidated into one of the other two.

**Comment by jdashg** [comment]
It looks to me like the title of this is resolved by #1341, so we should open/add to other issues anything we found here besides the title.
